lib/index.ts

```diff
@@ -454,6 +465,45 @@ export class DockerProgress {
 		return hash;
 	}
 
+	/** Build docker image calling onProgress with extended progress info regularly */
+	async build(
+		image: string,
+		onProgress: ProgressCallback,
+		options?: BuildOptions,
+	): Promise<string> {
+		// Create tar for build
+		const tarStream = tar.pack();
+		tarStream.entry({ name: 'Dockerfile' }, `FROM ${image}`, (err) => {
+			if (err) {
+				throw err;
+			}
+			tarStream.finalize();
+		});
+
+		const ignoreErrorEvents = !!options?.ignoreProgressErrorEvents;
+		const reporter = await this.getProgressReporter();
+		onProgress = reporter.pullProgress(onProgress);
+		const stream = await this.docker.buildImage(
+			tarStream,
+			options as ImageBuildOptions,
```

Comment:
Not a big deal, but would this type cast be avoided with the following?

Replace:
```ts
export interface BuildOptions extends PullPushOptions {
	t?: string; // Built image tag
	labels?: { [name: string]: string };
	keepParentImage?: boolean;
}
```
... with:
```ts
export interface BuildOptions extends PullPushOptions & Partial<ImageBuildOptions> {
	keepParentImage?: boolean;
}
```

It would also save two lines of code. :-) I didn't test this suggestion, so it may not work or not be that simple.

Reply:
I tried it, but the problem is that `authconfig` has a different format in `PushPullOptions` and `ImageBuildOptions`. For the first one, the type adds `auth` and `registrytoken` properties, but I'm not sure where those are used since they don't appear to be used in the repo (or in dockerode)

Reply:
I suppose that dockerode passes the extra options as query parameters to the local engine, in the case of balena devices, our engine understands the `auth` and `registrytoken` options

Reply:
@pdcastro what do you think, is ok to merge this?

Reply:
> the problem is that `authconfig` has a different format in `PushPullOptions` and `ImageBuildOptions`.
> I suppose that dockerode passes the extra options as query parameters to the local engine, in the case of balena devices, our engine understands the `auth` and `registrytoken` options

Well, this makes me even more worried than before, :-) now this could actually be a bug (not sure, but that's the point of having types). I am not aware of balenaEngine using a different `authconfig` format to the Docker engine.

I have tested that `npm build && npm test` for this repo succeed with the following changes that avoid the type cast:

```ts
export interface BuildOptions extends ImageBuildOptions {
	ignoreProgressErrorEvents?: boolean;
 }
```

```suggestion
			options,
```

After `npm run lint`, it actually becomes:

```ts
const stream = await this.docker.buildImage(tarStream, options);
```

How about doing it like that?

Reply:
Sounds good to me. I wonder if the compiler will complain when trying to use the module with the extra arguments, but I can figure that out if I run into that issue.

Reply:
Don't wonder, test it! To make testing easier, every time that the PR commits are updated, balena CI uploads a new version to the npm registry:

https://www.npmjs.com/package/docker-progress

Versions tab -> Version History section

Version | Downloads (Last 7 Days) | Published
--- | --- | ---
5.1.0-build-progress-95d538576857bacbab3128baa13e144f244d8105 | 0 | 28 minutes ago
5.1.0-build-progress-daf18688cbd96275471b27e70e945c6ac5a5fa7d | 28 | a day ago
...

In the downstream app, update `package.json` as follows:

Before: `"docker-progress": "^5.0.0",`
After: `"docker-progress": "5.1.0-build-progress-95d538576857bacbab3128baa13e144f244d8105",`

Then `npm i` and test the app with the new docker-progress.

After you have confirmed that everything works as expected, then you can change this PR from draft to non-draft and then ask reviewers to review it.

While at it, please confirm that `return hash;` returns the correct hash for the built image. Just wondering if the progress output for pull, push and build are really sufficiently compatible, such that `awaitRegistryStream` and `tryExtractDigestHash`, which were implemented for `push` and `pull`, work equally well for `build`.